packages/source-map/lib/sourceMap.ts

```diff
@@ -11,76 +11,151 @@ export interface Mapping<Data = unknown> {
-		const skip = new Set<Mapping>();
+		const startMatches: [Mapping<Data>, number][] = []
+		const startMatchesMap = new Map<Mapping<Data>, Set<number>>()
+		for (const [startMapping, index] of this.getStorageBasedOnRange(fromRange).findMatchingMappingIndexes(start, filter)) {
+			startMatches.push([startMapping, index])
+			let set = startMatchesMap.get(startMapping)
+			if (!set) {
+				set = new Set()
+				startMatchesMap.set(startMapping, set)
+			}
+			set.add(index)
+		}
 
-		for (let i = start; i <= end; i++) {
-			for (const mapping of memo.mappings[i]) {
-				if (skip.has(mapping)) {
-					continue;
+		const perfectMatches: [Mapping<Data>, number][] = []
+		const endMatches: [Mapping<Data>, number][] = []
+		for (const endMapping of this.getStorageBasedOnRange(fromRange).findMatchingMappingIndexes(start + length, filter)) {
+			const startMappingIndexes = startMatchesMap.get(endMapping[0])
+			if (startMappingIndexes && startMappingIndexes.has(endMapping[1])) {
+				perfectMatches.push(endMapping)
+			} else if (perfectMatches.length == 0) {
+				endMatches.push(endMapping)
+			}
+		}
+
+		const toRange: CodeRangeKey = fromRange == 'sourceOffsets' ? 'generatedOffsets' : 'sourceOffsets'
+		if (perfectMatches.length > 0) {
+			// Prefer the shortest mapping ranges
+			perfectMatches.sort((a, b) => getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const match of perfectMatches) {
+				const fromStartOffset = getStartOffset(match, fromRange)
+				const toStartOffset = getStartOffset(match, toRange)
+				const toLength = getSpanLength(match, toRange)
+				const fromLength = getSpanLength(match, fromRange)
+				if (toLength == fromLength) {
+					// `from` and `to` span mappings have the same length - map range verbatim
+					const startOffset = toStartOffset + start - fromStartOffset
+					yield [startOffset, startOffset + length, match[0]]
+				} else if (fromStartOffset == start && fromLength == length) {
+					// The whole `from` span is selected - map to the whole `to` span
+					yield [toStartOffset, toStartOffset + toLength, match[0]]
+				} else {
+					// We would need to do some heuristics here to map the span, and it would make little
+					// sense in terms of mapping accuracy.
+					// Try the next match.
 				}
-				skip.add(mapping);
+			}
+		} else {
+			// Prefer the shortest `from` spans
+			startMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+			endMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const startMatch of startMatches) {
+				let mapping = startMatch[0];
 
-				const mapped = translateOffset(offset, mapping[fromRange], mapping[toRange], getLengths(mapping, fromRange), getLengths(mapping, toRange));
-				if (mapped !== undefined) {
-					yield [mapped, mapping as Mapping<Data>] as const;
+				for (const endMatch of endMatches) {
+					if (endMatch[0] != mapping) {
```

Comment:
Not sure about this part, maybe we can have a second loop as a fallback, where we try to match just anything